### DISCUSSION TITLE
Hosted Service added in Module is not started

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#16231] [Q&A] [answered]
Hi,

I'm trying to run a background service in a module.

this line in the module startup is hit:

`            services.AddHostedService<KafkaConsumerService>();`

However, the background service never runs, what could be the issue?

I tried to start it manually, however the service isn't injected into the IOC

![image](https://github.com/OrchardCMS/OrchardCore/assets/25576022/d219e15c-21a9-47db-95c0-a168cd857283)

it only works if, I add it to startup as a singleton, then start in manually:
in module tenant start up events
`                await _provider.GetRequiredService<KafkaConsumerService>().StartAsync(stop);`

**Comment by lampersky** [comment, marked as answer]
@weirdyang HostedServices are only allowed at host level, you can read more [here](https://github.com/OrchardCMS/OrchardCore/issues/10103)

**Reply by weirdyang** [reply]
thanks I saw that, so what is the correct way to get the services related to the shell that is starting up or shutting down?

when I used IServiceProvider, it returned a new service, not the singleton i worked with earlier.

```
        public override Task TerminatingAsync()
        {
            return ShutdownAsync();
        }
        public async Task ShutdownAsync()
        {
            try
            {
                _provider.GetService<KafkaCancellationService>().TokenSource.Cancel();
                var stop = _provider.GetService<IHostApplicationLifetime>().ApplicationStopping;
                await _provider.GetRequiredService<KafkaConsumerService>().StopAsync(stop);
            }
            catch (Exception ex)
            {

                _logger.LogCritical(ex, "Unable to stop KafkaConsumerService"); ;
            }
        }
```

**Comment by weirdyang** [comment]
Ok so my current implementation is to:

1. hook into module tenant start up events
2. on start, start the service
3. on terminating, stop the service
4. using a cancellation token source linked to application lifetime

will there be any issue with this implementation?
